examples/post-sync/controller.go

```diff
@@ -0,0 +1,312 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+)
+
+// Sync kinds
+const (
+	RootSyncKind = "RootSync"
+	RepoSyncKind = "RepoSync"
+)
+
+// SyncID uniquely identifies a sync resource
+type SyncID struct {
+	Name      string
+	Kind      string
+	Namespace string
+}
+
+// StatusTracker tracks which errors have been logged
+type StatusTracker struct {
+	mu   sync.Mutex
+	seen map[string]struct{} // Using a single map with hashed keys
+}
+
+// NewStatusTracker creates a new StatusTracker instance
+func NewStatusTracker() *StatusTracker {
+	return &StatusTracker{
+		seen: make(map[string]struct{}),
+	}
+}
+
+// generateKey creates a unique key for a sync resource, commit, and error message
+func generateKey(syncID SyncID, commit, message string) string {
+	// Include SyncID information in the key
+	key := fmt.Sprintf("%s:%s:%s:%s:%s",
+		syncID.Name,
```

Comment:
nit: i would probably do kind:name:namespace:commit:msg, just because that's in order from most to least significant, but I don't know that it really matters unless you're printing them and need the list to be human scannable or sorted.

Reply:
The ordering doesn't matter much as the whole string is hashed, still changed for readability.